File: src/sort/QuickSort.java
package sort;


import java.util.Arrays;

public class QuickSort {

    static void quickSort(int[] items, int left, int right) {
        if (left < right) {
            int lastPivotPosition = partitionLeft(items, left, right);

            quickSort(items, left, lastPivotPosition - 1);
            quickSort(items, lastPivotPosition + 1, right);
        }
    }

    private static int partition(int[] items, int left, int right) {
        int i = left - 1;
        int pivot = items[right];

        for (int j = left; j < right; j++) {
            if (items[j] <= pivot) {
                swap(items, ++i, j);
            }
        }

        swap(items, ++i, right);

        return i;
    }

    private static int partitionLeft(int[] items, int left, int right) {
        int i = left - 1;
        int pivot = items[left];

        for (int j = left; j < right; j++) {
            if (items[j] <= pivot) {
                swap(items, ++i, j);
            }
        }

        swap(items, ++i, right);

        return i;
    }

    static void quickSort(int[] items) {
        quickSort(items, 0, items.length - 1);
    }

    private static void swap(int[] items, int source, int target) {
        int sourceTemp = items[source];
        items[source] = items[target];
        items[target] = sourceTemp;
    }

    public static void main(String[] args) {
        int[] items = new int[]{5, 8, 1, 3, 7, 9, 2};

        QuickSort.quickSort(items);

        for (int item : items) {
            System.out.print(item + " ");
        }
    }
}

File: src/sort/QuickSort.kt
fun quicksort(items: Array<Int>, left: Int, right: Int) {
    if (left < right) {
        val lastPivotPosition = partition(items, left, right)

        quicksort(items, left = left, right = lastPivotPosition - 1) // Left side
        quicksort(items, left = lastPivotPosition + 1, right = right) // Right side
    }
}

/**
 * @return last pivot position
 */
fun partition(items: Array<Int>, left: Int, right: Int): Int {
    val x = items[right]
    var i = left - 1

    for (j in left..right - 1) {
        if (items[j] <= x) {
            i++
            swap(items, i, j)
        }
    }

    i++
    swap(items, i, right)
    return i
}

fun swap(items: Array<Int>, i: Int, j: Int) {
    val itemI = items[i]
    items[i] = items[j]
    items[j] = itemI
}


fun main(args: Array<String>) {
    val array = arrayOf(5, 1, 2, 6, 9, 2, 2, 2220)

    quicksort(array, 0, array.size - 1)

    array.forEach { print(" $it") }
}